rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/responder.hpp

```diff
@@ -164,6 +164,49 @@ class Responder
     return estr;
   }
 
+  void deinit() noexcept
+  {
+    DDS::ReturnCode_t status;
+    if (response_datawriter_) {
+      status = response_publisher_->delete_datawriter(response_datawriter_);
+      if (nullptr != impl::check_delete_datawriter(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_datawriter(status));
+      }
+    }
+    if (response_topic_) {
+      status = participant_->delete_topic(response_topic_);
+      if (nullptr != impl::check_delete_topic(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_topic(status));
+      }
+    }
+    if (response_publisher_) {
+      status = participant_->delete_publisher(response_publisher_);
+      if (nullptr != impl::check_delete_publisher(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_publisher(status));
+      }
+    }
+    if (request_datareader_) {
+      // Assumption: subscriber is not null at this point.
+      status = request_subscriber_->delete_datareader(request_datareader_);
+      if (nullptr != impl::check_delete_datareader(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_datareader(status));
+      }
+    }
+    if (request_subscriber_) {
+      status = participant_->delete_subscriber(request_subscriber_);
+      if (nullptr != impl::check_delete_subscriber(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_subscriber(status));
+      }
+    }
+    if (request_topic_) {
+      status = participant_->delete_topic(request_topic_);
+      if (nullptr != impl::check_delete_topic(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_topic(status));
+      }
+    }
```

Comment:
If this is identical to the fail section of `init`, then I'd suggest making it a function and avoiding duplicate code.

However, it seems like the `deinit` function could return a string if there is there is a problem destructing things, similar to this:

https://github.com/ros2/rmw_opensplice/blob/master/rmw_opensplice_cpp/src/rmw_subscription.cpp#L242-L280

Basically you try each step (even if there are failures along the way) and if something fails you set the error string. But what's not captured clearly in the above link is that if the error string is already set (more than one thing failed) then the previous error string that's getting replaced is printed to stderr. So the `deinit` function could behave like this:

- Everything worked, return nullptr
- One thing failed, try all steps (don't return on failure) and return the error string for the thing that failed
- Multiple things failed, same as the previous point, but each time a new thing fails, print the error to stderr, and return the last error string through the normal channel

But I can tell you from experience that this is a rabbit hole of grace degradation that takes a lot of effort, so I think at some point just printing these problems to stderr is fine. The only exception to this is that the `deinit` function should probably cause the calling code to fail it it failed. This is the case for the `init` function: https://github.com/ros2/rmw_opensplice/pull/95/files#diff-cf8d0757343d0e3c757d5509dc29f18dR381